llvm/lib/CodeGen/CodeGenPrepare.cpp

```diff
@@ -8594,6 +8594,68 @@ static bool optimizeBranch(BranchInst *Branch, const TargetLowering &TLI,
   return false;
 }
 
+static bool tryNarrowMathIfNoOverflow(Instruction *I,
+                                      const TargetTransformInfo *TTI,
+                                      const DataLayout &DL) {
+  unsigned Opc = I->getOpcode();
+  Type *OldType = I->getType();
+
+  if (Opc != Instruction::Add && Opc != Instruction::Mul)
+    return false;
+
+  unsigned OrigBit = OldType->getScalarSizeInBits();
+  unsigned MaxBitsNeed = OrigBit;
+  switch (Opc) {
+  case Instruction::Add:
+    MaxBitsNeed = KnownBits::add(computeKnownBits(I->getOperand(0), DL),
+                                 computeKnownBits(I->getOperand(1), DL))
+                      .countMaxActiveBits();
+    break;
+  case Instruction::Mul:
+    MaxBitsNeed = KnownBits::mul(computeKnownBits(I->getOperand(0), DL),
+                                 computeKnownBits(I->getOperand(1), DL))
+                      .countMaxActiveBits();
+    break;
+  default:
+    break;
+  }
+
+  MaxBitsNeed = std::max<unsigned>(bit_ceil(MaxBitsNeed), 8);
```

Comment:
This should probably be rounding up to the next legal type? So you don't end up trying to create an i31 operation where i32 would do.